Update min/argmin and max/argmax to handle non-leading nans

jrbourbeau opened the pull request:
This PR adds `np.isnan` checks to ensure that we match the corresponding NumPy behavior. 

Partially addresses #4125. There is also a similar issue with datetime NaT values.

stuartarchibald commented:
Thanks very much for the PR, I've queued it for review.

jrbourbeau commented:
Great, thanks @stuartarchibald!

stuartarchibald commented on numba/targets/arraymath.py:
```diff
@@ -415,6 +415,22 @@ def array_min_impl(arry):
+                raise ValueError(MSG)
+
+            it = np.nditer(arry)
+            min_value = next(it).take(0)
```
If `arry[0]` is `NaN` such that `min_value` has the value `NaN` at this point the following loop need not proceed?

jrbourbeau replied:
Yes, good point. Returning at that point will help in the leading-`nan` case. I'll update accordingly

jrbourbeau commented:
@stuartarchibald I think I've addressed your comment about returning early when `arry[0]` is `NaN`. Please let me know if this is not the case, or if you have any additional comments

esc commented on numba/targets/arraymath.py:
```diff
@@ -415,6 +415,24 @@ def array_min_impl(arry):
                         min_value = v
             return min_value
 
+    elif isinstance(ty, types.Float):
```
The same maybe true for argmin/argmax, but there I can see the error message also differs.

stuartarchibald replied:
This is an excellent point in terms of code reuse, however it's probably best to close over the operator, passing functions as args leads to dispatch costs.

Consider:
```python
import numpy as np
from numba import njit
from IPython import get_ipython
ipython = get_ipython()
import operator

@njit
def lt(x, y):
    return operator.lt(x, y)

@njit
def op_arg(a, b, op):
    return op(a, b)

def gen_lt(op):
    def inner(a, b):
        op(a, b)
    return inner

close_arg = njit(gen_lt(operator.lt))

q, r = 1, 2

print("func arg : %s" % ipython.magic("timeit -o -q op_arg(q, r, lt)").best)
print("closure  : %s" % ipython.magic("timeit -o -q close_arg(q, r)").best)
```

jrbourbeau replied:
Thanks for the review @esc, good suggestion. IIUC it looks like closing over the operator is what's done for the existing `nanmin`/`nanmax` implementations: 

https://github.com/numba/numba/blob/ddd1cea985d61854bb1c07596a8c73c4f7f8468d/numba/targets/arraymath.py#L588-L662

We can try to take a similar approach here with the `min`/`max` implementations as well

jrbourbeau replied:
I ran into the issue of determining when a datetime value is "NaT" or not. Currently this is done by inspecting the `dtype` of `sig`

https://github.com/numba/numba/blob/ddd1cea985d61854bb1c07596a8c73c4f7f8468d/numba/targets/arraymath.py#L374-L381

This handles the datetime units correctly (e.g. `'datetime64[D]'` vs `'datetime64[Y]'`). Is there a way to get the proper datetime units without knowing the input array dtype? Naively using `nat = np.datetime64("NaT")` results in `TypingError`s like `Cannot unify datetime64[] and datetime64[Y] for 'min_value'`

stuartarchibald replied:
I think this probably needs `np.isnat` implementing as an internal function (and then it wiring into NumPy support with a version guard)?

stuartarchibald commented:
Just checking in on the state of this PR. @jrbourbeau was wondering if you have time to address the changes above, if not, don't worry, one of the core developers can take over to do the fixes. Thanks.
